### DISCUSSION TITLE
Crystal plasticity model

### DISCUSSION THREAD

**Discussion by Emmanuel-324** [idaholab/moose#27920] [Q&A Modules: Solid mechanics]
I am using the crystal plasticity kernel and the 'trial_xtalpl' as shown below is a required parameter of the crystal_plasticity_models object. I need more information about what 'trial_xtalpl' is but I can't find it. Is there a place to get more information on that, thanks.
@dschwen @vprithiv @GiudGiud 
[Materials]
  [stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
  []
[]

**Comment by GiudGiud** [comment]
Hello

Here it s just the name of another object in that input file I think.

Can you please paste the whole input file?

**Reply by Emmanuel-324** [reply]
```
[GlobalParams]
  displacements = 'ux uy uz'
[]

[Mesh]
  type = GeneratedMesh
  dim = 3
  elem_type = HEX8
[]

[AuxVariables]
  [fp_zz]
    order = CONSTANT
    family = MONOMIAL
  []
  [e_zz]
    order = CONSTANT
    family = MONOMIAL
  []
  [gss]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Functions]
  [tdisp]
    type = ParsedFunction
    expression = 0.01*t
  []
[]

[Physics/SolidMechanics/QuasiStatic/all]
  strain = FINITE
  add_variables = true
  generate_output = stress_zz
[]

[AuxKernels]
  [fp_zz]
    type = RankTwoAux
    variable = fp_zz
    rank_two_tensor = plastic_deformation_gradient
    index_j = 2
    index_i = 2
    execute_on = timestep_end
  []
  [e_zz]
    type = RankTwoAux
    variable = e_zz
    rank_two_tensor = total_lagrangian_strain
    index_j = 2
    index_i = 2
    execute_on = timestep_end
  []
  [gss1]
    type = MaterialStdVectorAux
    variable = gss
    property = slip_resistance
    index = 0
    execute_on = timestep_end
  []
[]

[BCs]
  [symmy]
    type = DirichletBC
    variable = uy
    boundary = bottom
    value = 0
  []
  [symmx]
    type = DirichletBC
    variable = ux
    boundary = left
    value = 0
  []
  [symmz]
    type = DirichletBC
    variable = uz
    boundary = back
    value = 0
  []
  [tdisp]
    type = FunctionDirichletBC
    variable = uz
    boundary = front
    function = tdisp
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeElasticityTensorCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
  []
  [stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
    maximum_substep_iteration = 200
    use_line_search = true
    min_line_search_step_size = 0.01
  []
  [trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    resistance_tol = 0.01
  []
[]

[Postprocessors]
  [stress_zz]
    type = ElementAverageValue
    variable = stress_zz
  []
  [fp_zz]
    type = ElementAverageValue
    variable = fp_zz
  []
  [e_zz]
    type = ElementAverageValue
    variable = e_zz
  []
  [gss]
    type = ElementAverageValue
    variable = gss
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  dt = 0.05
  solve_type = 'PJFNK'

  petsc_options_iname = -pc_hypre_type
  petsc_options_value = boomerang
  nl_abs_tol = 1e-10
  nl_rel_step_tol = 1e-10
  dtmax = 10.0
  nl_rel_tol = 1e-10
  end_time = 1
  dtmin = 0.02
  num_steps = 10
  nl_abs_step_tol = 1e-10
[]

[Outputs]
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
It's this object
```
  [trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    resistance_tol = 0.01
  []
```

**Reply by Emmanuel-324** [reply]
Does that mean I can choose any name I want for that block? Does trial_xtalpl do whatever [trial_xtalpl] block executes?

**Reply by GiudGiud** [reply]
yes. You can change that name to be whatever you want

**Comment by vprithiv** [comment]
@Emmanuel-324 

CrystalPlasticityKalidindiUpdate is the object we are dealing with, search in the codebase to see implementation. Please be aware that the model parameters have default values within this implementation. If you want to change that, you need to assign those parameters within the input file, like:

  [trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    **gss_initial = 120**
    slip_sys_file_name = input_slip_sys.txt
    resistance_tol = 0.01
  []

**Reply by Emmanuel-324** [reply]
Thanks @GiudGiud , @vprithiv . I am also looking for some information on the 'CrystalPlasticityStressUpdateBase' in the MOOSE documentation but I can't find it. Any help?

**Reply by vprithiv** [reply]
Check this one, most of the framework is in Compute Multiple Crystal Plasticity Stress 

https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html

CrystalPlasticityStressUpdateBase is the base class, I think it has the general framework. You can directly check the code in the codebase.
